### ISSUE TITLE
[Suggestion] Allow converting format with copyTextureToTexture

### ISSUE THREAD

**Issue by AshleyScirra** [gpuweb/gpuweb#1543]
tl;dr - converting a texture format on the GPU quickly becomes complicated. Currently copyTextureToTexture requires that the source and destination texture formats are the same. It would be convenient if it allowed changing the format, and handled the conversion steps internally.

---

Full rationale: I've been implementing an equivalent to WebGL readPixels in WebGPU, with the goal of reading back a specific render surface's pixel data as an ImageData (e.g. for saving as PNG, inspecting pixel data on CPU, etc.) It turns out ImageData uses data in RGBA8 format. However the preferred swapchain format on Windows is BGRA8, and so the surface texture format could be that, or in general any other texture format. So to read it back, it must first be converted to RGBA8 format, before using copyTextureToBuffer and mapping that buffer. For performance and avoiding having to write loads of format conversion code, this ought to be done on the GPU.

copyTextureToTexture cannot be used because it requires the same formats. Instead it can be done with a technique similar to mipmap generation: create a RENDER_ATTACHMENT texture, set up a pipeline and some shaders, and draw a quad. The destination texture can have a different format, providing the pipeline sets the correct target format. This all takes a fair amount of boilerplate code though. It also gets more complicated if you want to read back just a sub-rectangle of the render surface: now you have to work out some pixel-perfect co-ordinate conversions.

Also, this requires that the source texture have SAMPLED usage. If the texture has COPY_SRC usage but not SAMPLED, you can copy from it but not draw it this way. This can be worked around by doing another copy, to add the SAMPLED usage.

Further, the resulting converted texture has RENDER_ATTACHMENT usage. This doesn't seem to matter for copying to a buffer, but if in other cases you want to convert without adding this redundant usage, you'd have to add another copy to change the usage again. So in the end you can get a texture converted on the GPU, but it may end up taking 3 copies, and a lot of boilerplate code, all of which is a significant detour to the goal of reading back pixels.

Perhaps the intent of WebGPU is just to provide low-level primitives and let developers figure out all these details. But it would be considerably simpler if copyTextureToTexture allowed changing the format, and then have the browser work out all these complications. It would be significantly more convenient, and the browser might be able to specialise some paths to improve performance too. I guess it's a similar question to whether WebGPU should provide a convenience method for mipmap generation, or leave it up to developers. Perhaps the proposed image-to-image blit command in lieu of mipmaps as discussed in #386 could also cover this.

**Comment by kvark** [comment]
`copyTextureToTexture` has clear mapping to all the target APIs, where it is a byte-per-byte copy of the texture data. It can be used to re-interpret the data, but in no place it can actually do any conversions.

Therefore, if WebGPU implementations were to do the conversion, they'd have to basically go through all the jumps and hoops you described (creating the render pipelines, adding usage flags, issuing render/compute passes to copy, etc). Is there a reason this functionality should live inside a WebGPU implementation if you can do it on your end? Arguably, you can do it better than an implementation: you may know the format in advance, you may decide to copy from buffer to a texture in a compute shader, etc.

TL;DR: WebGPU implementation shouldn't be doing this

Note that there is "blit" functionality in *some* of the backends that you could use here (see #386), but exposing it is also controversial, and so far the group decided not to go there.

**Comment by AshleyScirra** [comment]
I guess that's fair enough, if it's a conscious choice that this is the philosophy of WebGPU then it makes sense.

One extra thing I'd point out though: it looks like GPUTexture does not have any JavaScript-accessible properties for usage, format, and the other creation parameters. This complicates writing library methods to handle things like texture format conversion which potentially needs a bunch of boilerplate code that is useful to move to a library. If you just pass a GPUTexture to some library "convertTextureFormat" method, then that method can't identify anything about that texture's format, usage, or size, so you'd have to pass all that in as additional parameters, and track them separately. This is a bit off-topic though, perhaps that could be filed as a different issue?

**Comment by kvark** [comment]
Right, this other issue you mentioned is discussed in #1498.
Closing this one, but please free to continue discussing and/or to re-open it if needed!
